File: core/src/com/ore/infinium/systems/InputSystem.java
package com.ore.infinium.systems;

import com.artemis.BaseSystem;
import com.artemis.ComponentMapper;
import com.artemis.annotations.Wire;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.ore.infinium.OreWorld;
import com.ore.infinium.components.*;
import com.ore.infinium.systems.profiler.SystemProfiler;

/**
 * ***************************************************************************
 * Copyright (C) 2015 by Shaun Reich <sreich02@gmail.com>
 * <p>
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 * <p>
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * <p>
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * ***************************************************************************
 */
@Wire
/**
 * This input system is stuck in a bit of a hairy position...for input, we need two
 * scenarios, direct input event, and polling based. Polling based must be used for things like
 * primary attack.
 *
 * Event based is used for pretty much everything else, like GUI related things (excluding scene2d, that's already
 * basically handled by itself).
 */ public class InputSystem extends BaseSystem {
    private OrthographicCamera m_camera;

    private ComponentMapper<PlayerComponent> playerMapper;
    private ComponentMapper<SpriteComponent> spriteMapper;
    private ComponentMapper<ControllableComponent> controlMapper;
    private ComponentMapper<ItemComponent> itemMapper;
    private ComponentMapper<VelocityComponent> velocityMapper;
    private ComponentMapper<JumpComponent> jumpMapper;

    private PowerOverlayRenderSystem m_powerOverlayRenderSystem;

    private OreWorld m_world;

    public InputSystem(OrthographicCamera camera, OreWorld world) {
        m_camera = camera;
        m_world = world;

    }

    @Override
    protected void processSystem() {
        //ignore input if the profiler is shown
        if (SystemProfiler.isRunning()) {
            return;
        }

        if (m_world.m_client.leftMouseDown && !m_powerOverlayRenderSystem.overlayVisible) {

            m_world.m_client.handleLeftMousePrimaryAttack();
        }
    }

}

File: core/src/com/ore/infinium/systems/client/InputSystem.kt
/**
MIT License

Copyright (c) 2016 Shaun Reich <sreich02@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */

package com.ore.infinium.systems.client

import com.artemis.BaseSystem
import com.artemis.annotations.Wire
import com.badlogic.gdx.graphics.OrthographicCamera
import com.ore.infinium.OreWorld
import com.ore.infinium.util.system

@Wire
/**
 * This input system is stuck in a bit of a hairy position...for input, we need two
 * scenarios, direct input event, and polling based. Polling based must be used for things like
 * primary attack.

 * Event based is used for pretty much everything else, like GUI related things (excluding scene2d, that's already
 * basically handled by itself).
 */
class InputSystem(private val camera: OrthographicCamera, private val oreWorld: OreWorld) : BaseSystem() {

    private val powerOverlayRenderSystem by system<PowerOverlayRenderSystem>()
    private val clientNetworkSystem by system<ClientNetworkSystem>()

    override fun processSystem() {
        if (powerOverlayRenderSystem.overlayVisible || !clientNetworkSystem.connected) {
            return
        }

        if (oreWorld.client!!.leftMouseDown) {
            oreWorld.client!!.handlePrimaryAttack()
        } else if (oreWorld.client!!.rightMouseDown) {
            oreWorld.client!!.handleSecondaryAttack()
        }
    }
}
